File: Sort/C++/radixsort.cpp
#include "main.h"

/**
 * getmax - gets max in an array
 *
 * @arr: array
 * @n: size of array
 *
 * Return: max
 */
int getmax(int arr[], int n)
{
	int max = arr[0];
	int i;

	for (i = 1; i < n; i++)
	{
		if (arr[i] > max)
			max = arr[i];
	}
	return (max);
}

/**
 * countingsort - sorting technique
 * @arr: array
 * @n: size of array
 * @pos: place value
 *
 * Return: void
 */
void countingsort(int arr[], int n, int pos)
{
	int i;
	int count[10], output[n];

	for (i = 0; i < 10; i++)
		count[i] = 0;

	for (i = 0; i < n; i++)
		++count[(arr[i] / pos) % 10];

	for (i = 1; i < 10; i++)
		count[i] += count[i - 1];

	for (i = 0; i < n; i++)
		output[--count[(arr[i] / pos) % 10]] = arr[i];

	for ( i = 0; i < n; i++)
		arr[i] = output[i];
}

/**
 * radixsort - sorting technique
 *
 * @arr: array
 * @n: size of array
 *
 * Return: void
 */
void radixsort(int arr[], int n)
{
	int max = getmax(arr, n);
	int pos;

	for(pos = 1; max / pos > 0; pos *= 10)
		countingsort(arr, n , pos);
}

File: Sort/Java/RadixSort.java
public class RadixSort {
    public static class Sort{
        public int max(int[] arr){
            int m = arr[0];
            for (int i = 1; i < arr.length; i++){
                if (arr[i] > m){
                    m = arr[i];
                }
            }
            return m;
        }
        public void countingsort(int[] arr, int pos){
            int n = arr.length;
            int count[] = new int[10];
            int output[] = new int[n];

            for (int i = 0; i < count.length; i++) {
                count[i] = 0;
            }
            for (int i = 0; i < n; i++) {
                ++count[(arr[i] / pos) % 10];
            }
            for (int i = 1; i < count.length; i++) {
                count[i] += count[i - 1];
            }
            for (int i = 0; i < n; i++) {
                output[--count[(arr[i] / pos) % 10]] = arr[i];
            }
            for (int j = 0; j < n; j++) {
                arr[j] = output[j];
            }
        }
        public void radixsort(int[] arr){
            int max = max(arr);
            for (int pos = 1; max / pos > 0; pos *= 10){
                countingsort(arr, pos);
            }
        }
    }
    public static void printarray(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }
    public static void main(String[] args) {
        int[] arr = {93, 88, 77, 26, 55, 63, 32, 40, 01};

        Sort obj = new Sort();
        System.out.println("Array before sorted: ");
        printarray(arr);
        obj.radixsort(arr);
        System.out.println("Array after sorted: ");
        printarray(arr);
    }
}
